packages/structures/src/users/User.ts

```diff
@@ -0,0 +1,193 @@
+import { DiscordSnowflake } from '@sapphire/snowflake';
+import type { APIUser } from 'discord-api-types/v10';
+import { Structure } from '../Structure.js';
+import { data as kData } from '../utils/symbols.js';
+
+let UserDataTemplate: Partial<APIUser> = {};
+
+/**
+ * Sets the template used for removing data from the raw data stored for each User
+ *
+ * @param template - the template
+ */
+export function setUserDataTemplate(template: Partial<APIUser>) {
+	UserDataTemplate = template;
+}
+
+/**
+ * Gets the template used for removing data from the raw data stored for each User
+ *
+ * @returns the template
+ */
+export function getUserDataTemplate(): Readonly<Partial<APIUser>> {
+	return UserDataTemplate;
+}
+
+/**
+ * Represents any user on Discord.
+ */
+export class User<Omitted extends keyof APIUser | '' = ''> extends Structure<APIUser> {
+	public constructor(
+		/**
+		 * The raw data received from the API for the user
+		 */
+		data: Omit<APIUser, Omitted>,
+	) {
+		// Cast here so the getters can access the properties, and provide typesafety by explicitly assigning return values
+		super(data as APIUser, { template: UserDataTemplate });
+	}
+
+	public override patch(data: Partial<APIUser>) {
+		return super.patch(data, { template: UserDataTemplate });
+	}
+
+	/**
+	 * The user's id
+	 */
+	public get id() {
+		return this[kData].id as 'id' extends Omitted ? never : APIUser['id'];
```

Comment:
If we'll have to add this return type for each getter
1. let's add it as the return type, if we can, else keep it as a cast
2. Let's make a utility type for this (`NeverIfOmitted<Omitted, 'id', APIUser['id']>`)

Reply:
Definitely utility type. I wish it didn't have to cast, but [value of thing] is not assignable to never (or undefined). Maybe we can add `| never` to all the types in `Structure[kData]` though and then it works? I'll try that, cause it'd make it a lot easier if it were a merged type.

Reply:
Unfortunately not no. One, it's not as nice of an interface for consumers. Two, this doesn't allow us to convert from snake_case to camelCase.

I would love to do it this way, since it's a lot easier to maintain for us, but I think prioritizing the consumer interface is more important here. These base structures are a lot to implement at first, but shouldn't need to be touched super often once implemented.